### PR TITLE
Tfvars lookup in workspace

### PR DESCRIPTION
**Pull request by opteemister**
It would be great to have separate tfvars files for separate workspaces/environments.
I think it could be enough for lookup in workspace folder too.
Please feel free to improve the code it there are any existing parts of code with the same functionality.

### CODE REVIEW

**Comment by opteemister** [comment]
@stack72 please look at it. Thank you)

**Comment by rowleyaj** [comment]
Could this lead people to use workspaces against the best practices documented in the [workspaces doc](https://github.com/hashicorp/terraform/blob/master/website/docs/state/workspaces.html.md) by making it easier to use workspaces as the only isolation?

>Workspaces can be used to manage small differences between development, staging, and production, but they should not be treated as the only isolation mechanism.

In the past using older versions of Terraform I've had separate folders for environments dev/stage/prod etc, and then subfolders within for stacks and used remote state and modules to compose the infrastructure for the environment. This isolated the variables for each environment, and the state preventing any cross-environment issues. @nickithewatt described a similar workflow in her [talk at HashiDays London](https://twitter.com/techiewatt/status/874579034507804673)

Is the intended workflow with this PR and workspaces to move away from folders isolating environments and instead use only workspaces for isolation - although of much smaller sections of infrastructure?

#14869 looked to clarify the usage a little but i'm not sure if it could go further, maybe as part of this PR to introduce per workspace variables?

File: command/meta.go [hunk 1]
```diff
@@ -483,6 +499,15 @@ func (m *Meta) outputShadowError(err error, output bool) bool {
 	return true
 }
 
+func (m *Meta) EnvPath() string {
+  envName := m.Workspace()
```
Comment (rowleyaj) [inline command/meta.go]: `workspaceName` ?

File: command/meta.go [hunk 2]
```diff
@@ -483,6 +499,15 @@ func (m *Meta) outputShadowError(err error, output bool) bool {
 	return true
 }
 
+func (m *Meta) EnvPath() string {
```
Comment (rowleyaj) [inline command/meta.go]: `WorkspacePath` ?

File: command/meta.go [hunk 3]
```diff
@@ -483,6 +499,15 @@ func (m *Meta) outputShadowError(err error, output bool) bool {
...
+  envName := m.Workspace()
+	if envName == backend.DefaultStateName {
+		return ""
+	} else {
+		return filepath.Join(local.stateWorkspaceDir(), envName)
```
Comment (rowleyaj) [inline command/meta.go]: Would this be the right place to look for these variable files? My understanding is that this directory would only exist when using local state?

Could this look for the variable files alongside the .tf files but within folders named after the workspace?

**Comment by apparentlymart** [comment]
Hi @opteemister! Thanks for working on this.

Having a per-workspace variables file (or other ways to achieve the same thing) is something we've been talking about for a while and planning to do. We didn't start working deeply on it yet because there are some touch points with the Terraform Enterprise product that we need to figure out, but this patch looks like a good start!

I'd like to ask for some patience here was we do some design work internally to make sure this will align with the enterprise product plans, since we want to avoid painting ourselves into a corner. I'll circle back here once we have a more definite idea of what the deal is.

**Comment by opteemister** [comment]
Thank you for looking.
@rowleyaj you are right from one side. But in some cases it will be very useful to have per-workspace variables. I just describe one use-case which we have on our project:
1. We have 2 environments (staging, prod) which should be exactly the same (but with small differences in such things like passwords, ip ranges, sometimes different instance sizes) but in total it should be the same configuration (count of instances, DB cluster configuration, network configuration, etc.)
2. Let's imagine that we follow the Best Practices. And we're creating network infrastructure for these environments. We need to have the same topology but with different ip ranges (e.g. 172.0.0.0/16 and 172.1.0.0/16) that would be defined in per-workspace tfvars. And in that case we don't have to duplicate the configuration or create Maps with defining IP range per-env.
3. Then we can create DNS records (Route53). And we will have the same subdomains, but different main domain. In that case we will follow Best Practices and use network as a remote-state and create DNS records based on it. But again it would be great to have different per-env variables.
4. And there should be also a common variable file to store common variables.

We use that schema for a long time and didn't have any problems with that. Here is an example how we call terraform without using environments/workspaces (of course in script with variables):
`terraform plan -var-file=common.tfvars -var-file=environment.tfvars -state=environment.tfstate`
But I hope to use native solution to remove such workaround.

And one more thing. I saw that it will be possible to use TF_WORKSPACE ENV_VAR. That is awesome. 
In case when I have script for exporting ENV_VAR for (e.g. AWS) provider I can define this variable there and use terraform without additional switching workspace before use.

**Comment by opteemister** [comment]
@apparentlymart 
Thank you for the explanation. I hope you will come back soon and we can use something like that in 0.1x versions.

**Comment by teamterraform** [comment]
Hello, and sorry for the long silence!
There is an ongoing internal discussion about the design of workspaces which will likely change much about how workspaces work, we are going to close this PR. Thank you!

**Pull request closed without merging**